File: packages/algo-py/solutions/algo_02.py
class TreeNode:
    def __init__(self, val):
        self.left = None
        self.right = None
        self.val = val


def is_valid_BST(node, min, max):
    if node == None:
        return True
    if (min is not None and node.val <= min) or (max is not None and max <= node.val):
        return False
    return is_valid_BST(node.left, min, node.val) and is_valid_BST(node.right, node.val, max)

File: packages/algo-cpp/solutions/algo_02.cpp
#include <gtest/gtest.h>
#include <stdio.h>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;

  TreeNode(int val) : val(val), left(NULL), right(NULL) {}
};

bool is_valid_BST(TreeNode *node, int *max, int *min)
{
  if (node == NULL)
  {
    return true;
  }
  if ((max != NULL && *max <= node->val) || (min != NULL && node->val <= *min))
  {
    return false;
  }

  return is_valid_BST(node->left, &node->val, min) && is_valid_BST(node->right, max, &node->val);
}

TEST(IsValidBST, TestValid)
{
  TreeNode root(5);
  TreeNode leftChild(4);
  TreeNode rightChild(6);

  root.left = &leftChild;
  root.right = &rightChild;

  EXPECT_TRUE(is_valid_BST(&root, NULL, NULL));
}

TEST(IsValidBST, TestInvalid)
{
  TreeNode root(5);
  TreeNode leftChild(4);
  TreeNode rightChild(6);

  root.right = &leftChild;
  root.left = &rightChild;

  EXPECT_FALSE(is_valid_BST(&root, NULL, NULL));
}
